From 1cf71f385888011ca46781d641ecad8276931d4d Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Sun, 12 Nov 2023 12:09:18 +0100
Subject: [PATCH 140/174] net: dovetail: add missing argument to
 sock_oob_connect()

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 net/socket.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/net/socket.c b/net/socket.c
index 28752faced82..b005fc497e40 100644
--- a/net/socket.c
+++ b/net/socket.c
@@ -167,7 +167,8 @@ int __weak sock_oob_bind(struct socket *sock, struct sockaddr *addr, int len)
 	return 0;
 }
 
-int __weak sock_oob_connect(struct socket *sock, struct sockaddr *addr, int len)
+int __weak sock_oob_connect(struct socket *sock,
+			struct sockaddr *addr, int len, int flags)
 {
 	return 0;
 }
@@ -223,7 +224,7 @@ static inline int sock_oob_bind(struct socket *sock,
 }
 
 static inline int sock_oob_connect(struct socket *sock,
-				struct sockaddr *addr, int len)
+				struct sockaddr *addr, int len, int flags)
 {
 	return 0;
 }
@@ -2177,7 +2178,7 @@ int __sys_connect_file(struct file *file, struct sockaddr_storage *address,
 				addrlen, sock->file->f_flags | file_flags);
 	if (!err && sock_oob_capable(sock)) {
 		err = sock_oob_connect(sock, (struct sockaddr *)address,
-				addrlen);
+				addrlen, sock->file->f_flags | file_flags);
 		if (err)
 			goto out;
 	}
-- 
2.43.0

