From 19f806d707e85955dd554626445a34209b5327a4 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Thu, 4 Apr 2019 12:29:11 +0200
Subject: [PATCH 035/171] gpio: omap: irq_pipeline: enable pipelined interrupt
 control

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 drivers/gpio/gpio-omap.c | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/drivers/gpio/gpio-omap.c b/drivers/gpio/gpio-omap.c
index a927680c66f8..7f42b4e40da6 100644
--- a/drivers/gpio/gpio-omap.c
+++ b/drivers/gpio/gpio-omap.c
@@ -56,8 +56,8 @@ struct gpio_bank {
 	u32 saved_datain;
 	u32 level_mask;
 	u32 toggle_mask;
-	raw_spinlock_t lock;
-	raw_spinlock_t wa_lock;
+	hard_spinlock_t lock;
+	hard_spinlock_t wa_lock;
 	struct gpio_chip chip;
 	struct clk *dbck;
 	struct notifier_block nb;
@@ -728,7 +728,7 @@ static const struct irq_chip omap_gpio_irq_chip = {
 	.irq_bus_lock = omap_gpio_irq_bus_lock,
 	.irq_bus_sync_unlock = gpio_irq_bus_sync_unlock,
 	.irq_print_chip = omap_gpio_irq_print_chip,
-	.flags = IRQCHIP_MASK_ON_SUSPEND | IRQCHIP_IMMUTABLE,
+	.flags = IRQCHIP_MASK_ON_SUSPEND | IRQCHIP_IMMUTABLE | IRQCHIP_PIPELINE_SAFE,
 	 GPIOCHIP_IRQ_RESOURCE_HELPERS,
 };
 
@@ -741,7 +741,7 @@ static const struct irq_chip omap_gpio_irq_chip_nowake = {
 	.irq_bus_lock = omap_gpio_irq_bus_lock,
 	.irq_bus_sync_unlock = gpio_irq_bus_sync_unlock,
 	.irq_print_chip = omap_gpio_irq_print_chip,
-	.flags = IRQCHIP_MASK_ON_SUSPEND | IRQCHIP_IMMUTABLE,
+	.flags = IRQCHIP_MASK_ON_SUSPEND | IRQCHIP_IMMUTABLE | IRQCHIP_PIPELINE_SAFE,
 	 GPIOCHIP_IRQ_RESOURCE_HELPERS,
 };
 
@@ -1077,7 +1077,7 @@ static int omap_gpio_chip_init(struct gpio_bank *bank, struct device *pm_dev)
 	irq_domain_set_pm_device(bank->chip.irq.domain, pm_dev);
 	ret = devm_request_irq(bank->chip.parent, bank->irq,
 			       omap_gpio_irq_handler,
-			       0, dev_name(bank->chip.parent), bank);
+			       IRQF_OOB, dev_name(bank->chip.parent), bank);
 	if (ret)
 		gpiochip_remove(&bank->chip);
 
-- 
2.43.0

