From 016f5ad961fe88da9262758406149d57fe60cd6c Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Thu, 21 Sep 2023 09:33:31 +0200
Subject: [PATCH 102/171] dovetail: fix build w/ IRQ_PIPELINE + !DOVETAIL

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 include/linux/dovetail.h | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

diff --git a/include/linux/dovetail.h b/include/linux/dovetail.h
index 70b78fd55d5e..366260e3291c 100644
--- a/include/linux/dovetail.h
+++ b/include/linux/dovetail.h
@@ -6,6 +6,12 @@
 #ifndef _LINUX_DOVETAIL_H
 #define _LINUX_DOVETAIL_H
 
+struct pt_regs;
+struct mm_struct;
+struct task_struct;
+struct file;
+struct files_struct;
+
 #ifdef CONFIG_DOVETAIL
 
 #include <linux/sched.h>
@@ -15,11 +21,6 @@
 #include <uapi/asm-generic/dovetail.h>
 #include <asm/dovetail.h>
 
-struct pt_regs;
-struct task_struct;
-struct file;
-struct files_struct;
-
 enum inband_event_type {
 	INBAND_TASK_SIGNAL,
 	INBAND_TASK_MIGRATION,
@@ -230,13 +231,13 @@ void replace_inband_fd(unsigned int fd, struct file *file,
 
 #else	/* !CONFIG_DOVETAIL */
 
+#include <linux/irqflags.h>
+
 #ifdef CONFIG_HAVE_DOVETAIL
 /* We may have arch-specific placeholders. */
 #include <asm/dovetail.h>
 #endif
 
-struct files_struct;
-
 #define protect_inband_mm(__flags)	\
 	do { (void)(__flags); } while (0)
 
-- 
2.43.0

