From 80b32a1d9861f13b90af87ffc6033182c8f40dda Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Thu, 22 Feb 2024 14:45:57 +0100
Subject: [PATCH 136/171] sched: irq_pipeline: make migration check oob-aware

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 kernel/sched/core.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index 9c345f7c02c9..66c542f5b018 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -10326,6 +10326,9 @@ void __cant_migrate(const char *file, int line)
 {
 	static unsigned long prev_jiffy;
 
+	if (irqs_pipelined() && (running_oob() || hard_irqs_disabled()))
+		return;
+
 	if (irqs_disabled())
 		return;
 
-- 
2.43.0

