From 47464413f4a185def3fca9eb81f848e7d2a2572c Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Fri, 1 Mar 2024 10:49:27 +0100
Subject: [PATCH 147/171] net: dovetail: document ndo out-of-band handlers

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 include/linux/netdevice.h | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/include/linux/netdevice.h b/include/linux/netdevice.h
index 9740337df261..0ac9a2246c32 100644
--- a/include/linux/netdevice.h
+++ b/include/linux/netdevice.h
@@ -1408,6 +1408,18 @@ struct netdev_net_notifier {
  *			   struct kernel_hwtstamp_config *kernel_config,
  *			   struct netlink_ext_ack *extack);
  *	Change the hardware timestamping parameters for NIC device.
+ * struct sk_buff *(*ndo_alloc_oob_skb)(struct net_device *dev,
+ * 					dma_addr_t *dma_addr);
+ *	Allocate a buffer for out-of-band I/O. The memory must be immediately
+ *	DMA-suitable, living in the linear kernel address space. i.e. no highmem.
+ *	This handler must be provided if IFF_OOB_CAPABLE is set in the out-of-band
+ *	context flags.
+ * void	(*ndo_free_oob_skb)(struct net_device *dev,
+ *			    struct sk_buff *skb,
+ *			    dma_addr_t dma_addr);
+ *	Free a buffer previously allocated from ndo_alloc_oob_skb() exclusively.
+ *	This handler must be provided if IFF_OOB_CAPABLE is set in the out-of-band
+ *	context flags.
  */
 struct net_device_ops {
 	int			(*ndo_init)(struct net_device *dev);
@@ -4312,6 +4324,13 @@ void netif_device_attach(struct net_device *dev);
 
 bool netif_receive_oob(struct sk_buff *skb);
 
+/**
+ *	netif_oob_diversion - is I/O traffic diverted for out-of-band handling.
+ *	@dev: network device
+ *
+ * Check if the RX/TX packets are currently diverted for out-of-band
+ * handling by a companion kernel.
+ */
 static inline bool netif_oob_diversion(const struct net_device *dev)
 {
 	return test_bit(__LINK_STATE_OOB, &dev->state);
-- 
2.43.0

